util/src/main/java/io/kubernetes/client/util/ClientBuilder.java

```diff
@@ -12,231 +12,217 @@
+    if(kubeConfigReader != null) {
+      return kubeconfig(loadKubeConfig(kubeConfigReader));
+    }
+    final FileReader configReader = findConfigInHomeDir();
+    if(configReader != null) {
+      return kubeconfig(loadKubeConfig(configReader));
+    }
+    final File clusterCa = new File(SERVICEACCOUNT_CA_PATH);
+    if (clusterCa.exists()) {
+      return cluster();
+    }
+    return new ClientBuilder();
+  }
+
+  private static FileReader findConfigFromEnv() {
+    try {
+      String kubeConfig = System.getenv(ENV_KUBECONFIG);
+      if(kubeConfig == null) {
+        return null;
+      }
+      return new FileReader(kubeConfig);
+    } catch (FileNotFoundException e) {
+      log.info("Could not find file specified in $KUBECONFIG");
+      return null;
+    }
+  }
+
+  private static FileReader findConfigInHomeDir() {
+    try {
+      File config = new File(new File(System.getenv(ENV_HOME), KUBEDIR), KUBECONFIG);
+      return new FileReader(config);
+    } catch (FileNotFoundException e) {
+      log.info("Could not find ~/.kube/config");
+      return null;
+    }
+  }
+
+  /**
+   * Creates a builder which is pre-configured from the cluster configuration.
+   *
+   * @return <tt>ClientBuilder</tt> configured from the cluster configuration.
+   * @throws IOException
+   *  if the Service Account Token Path or CA Path is not readable.
+   */
+  public static ClientBuilder cluster() throws IOException {
+    final ClientBuilder builder = new ClientBuilder();
+
+    final String host = System.getenv(ENV_SERVICE_HOST);
+    final String port = System.getenv(ENV_SERVICE_PORT);
+    builder.setBasePath("https://" + host + ":" + port);
+
+    final String token = new String(Files.readAllBytes(Paths.get(SERVICEACCOUNT_TOKEN_PATH)),
+        Charset.defaultCharset());
+    builder.setCertificateAuthority(Files.readAllBytes(Paths.get(SERVICEACCOUNT_CA_PATH)));
+    builder.setAuthentication(new AccessTokenAuthentication(token));
+
+    return builder;
+  }
+
+  /**
+   * Creates a builder which is pre-configured from a {@link KubeConfig}.
+   *
+   * To load a <tt>KubeConfig</tt>, see {@link KubeConfig#loadKubeConfig(Reader)}.
+   *
+   * @param config
+   *  The {@link KubeConfig} to configure the builder from.
+   * @return <tt>ClientBuilder</tt> configured from the provided <tt>KubeConfig</tt>
+   * @throws IOException
+   *  if the files specified in the provided <tt>KubeConfig</tt> are not readable
+   */
+  public static ClientBuilder kubeconfig(KubeConfig config) throws IOException {
+    final ClientBuilder builder = new ClientBuilder();
+
+    String server = config.getServer();
+    if (!server.startsWith("http://") && !server.startsWith("https://")) {
+      if (server.contains(":443")) {
+        server = "https://" + server;
+      } else {
+        server = "http://" + server;
+      }
+    }
+
+    if(config.verifySSL()) {
+      final byte[] caBytes = KubeConfig.getDataOrFile(config.getCertificateAuthorityData(),
```

Comment:
This isn't required for `verifySSL` imagine if someone has a cluster with a valid globally issued CA, they may not need to load in a custom CA...

Reply:
Added check to see if `caBytes` is null before making this call. Force `setVerifyingSsl` to true on the next line.

Reply:
This logic feels convoluted to me.

Why not:
```java
byte[] caBytes = ...
if (caBytes != null) {
   builder.setCertificateAuthority(caBytes);
}
builder.setVerifyingSsl(true);
```

That feels cleaner to me.

Yes, it's unlikely that someone would set verifySSL to false and still pass a ca authority, but I don't think there's any harm in allowing it...

Reply:
You are right. I will remove the `if` statement related to `config.verifySSL` and also make `setCertificateAuthority` no longer set `verifyingSsl` to `true` as that is the default value.

This should provide the following behaviour
- Having `certificate-authority` or `certificate-authority-file` set will make the client verify with provided CA cert.
- Having `insecure-skip-tls-verify` set to `false` will set `verifyingSsl` to `false` regardless of providing a CA cert above.